packages/expo-screen-orientation/ios/EXScreenOrientation/EXScreenOrientationRegistry.m

```diff
@@ -0,0 +1,227 @@
+  }
+  return currentOrientationMask;
+}
+
+- (UIInterfaceOrientation)currentOrientation
+{
+  return _currentScreenOrientation;
+}
+
+#pragma mark - events
+
+- (void)handleDeviceOrientationChange:(NSNotification *)notification
+{
+  UIInterfaceOrientation newScreenOrientation = [EXScreenOrientationUtilities UIDeviceOrientationToUIInterfaceOrientation:[notification.object orientation]];
+  [self deviceOrientationDidChange:newScreenOrientation];
+}
+
+- (void)deviceOrientationDidChange:(UIInterfaceOrientation)newScreenOrientation
+{
+  if (_currentScreenOrientation == newScreenOrientation || newScreenOrientation == UIInterfaceOrientationUnknown) {
+    return;
+  }
+  
+  UIInterfaceOrientationMask currentOrientationMask = [self currentOrientationMask];
+  
+  // checks if screen orientation should be changed when user rotates the device
+  if ([EXScreenOrientationUtilities doesOrientationMask:currentOrientationMask containOrientation:newScreenOrientation]) {
+    // change current screen orientation
+    if ((UIInterfaceOrientationIsPortrait(newScreenOrientation) && UIInterfaceOrientationIsPortrait(_currentScreenOrientation))
+        || (UIInterfaceOrientationIsLandscape(newScreenOrientation) && UIInterfaceOrientationIsLandscape(_currentScreenOrientation))) {
+      _currentScreenOrientation = newScreenOrientation; // updates current screen orientation, but doesn't emit event
+      return;
+    }
+    
+    // on iPads, traitCollectionDidChange isn't triggered at all
+    if ([EXScreenOrientationUtilities isIPad] &&
+        ((UIInterfaceOrientationIsPortrait(newScreenOrientation) && UIInterfaceOrientationIsLandscape(_currentScreenOrientation))
+         || (UIInterfaceOrientationIsLandscape(newScreenOrientation) && UIInterfaceOrientationIsPortrait(_currentScreenOrientation)))) {
+      [self screenOrientationDidChange:newScreenOrientation];
+    }
+  }
+}
+
+- (void)traitCollectionsDidChange:(UITraitCollection *)traitCollection
+{
+  UIUserInterfaceSizeClass verticalSizeClass = traitCollection.verticalSizeClass;
+  UIUserInterfaceSizeClass horizontalSizeClass = traitCollection.horizontalSizeClass;
+  UIInterfaceOrientation currentDeviceOrientation = [EXScreenOrientationUtilities UIDeviceOrientationToUIInterfaceOrientation:[[UIDevice currentDevice]  orientation]];
+  UIInterfaceOrientationMask currentOrientationMask = [[[[UIApplication sharedApplication] keyWindow] rootViewController] supportedInterfaceOrientations];
+   
+  UIInterfaceOrientation newScreenOrientation = UIInterfaceOrientationUnknown;
+
+  if (verticalSizeClass == UIUserInterfaceSizeClassRegular && horizontalSizeClass == UIUserInterfaceSizeClassCompact) {
+    // portrait
+    UIInterfaceOrientationMask portraitMask = currentOrientationMask & UIInterfaceOrientationPortrait;
+    if (portraitMask == UIInterfaceOrientationMaskPortrait) {
+      newScreenOrientation = UIInterfaceOrientationPortrait;
+    } else if (portraitMask == UIInterfaceOrientationMaskPortraitUpsideDown) {
```

Comment:
Will this clause ever be true? If `portraitMask = … & UIInterfaceOrientationPortrait` I doubt it that it can ever be equal to something other than `UIInterfaceOrientationMaskPortrait`, but I may be wrong

Reply:
`MaskPortrait` isn't in the same relationship with `OrientationPortrait` and `OrientationPortraitUpsideDown` as `MaskLandscape` is with `OrientationLandscapeLeft` and `OrientationLandscapeRight`

```objc
    UIInterfaceOrientationMaskPortrait = (1 << UIInterfaceOrientationPortrait),
    UIInterfaceOrientationMaskPortraitUpsideDown = (1 << UIInterfaceOrientationPortraitUpsideDown),

    UIInterfaceOrientationMaskLandscapeLeft = (1 << UIInterfaceOrientationLandscapeLeft),
    UIInterfaceOrientationMaskLandscapeRight = (1 << UIInterfaceOrientationLandscapeRight),
    UIInterfaceOrientationMaskLandscape = (UIInterfaceOrientationMaskLandscapeLeft | UIInterfaceOrientationMaskLandscapeRight),
```
so what we could use in landscape case won't work here